### DISCUSSION TITLE
Compatibility with Cuda

### DISCUSSION THREAD

**Discussion by madzai** [deepmind/open_spiel#1036] [Q&A] [answered]
I have installed Cuda 12.1 in Ubuntu Linux to enable GPU to run OpenSpiel training of game agents. Some error appeared when running build_and_run_tests.sh. 

OpenSpiel was compiled OK before Cuda was installed. The CMake version on my machine is 3.22.1. Can I ask if it may be due to any known compatibility issue with Cuda?

_______________________________________________________________________________________________________________________________________

Below are part of the results:

Building and testing in /home/user/Documents/OpenSpiel folders/open_spiel/build using 'python' (version 3.11.2).
-- The C compiler identification is GNU 11.3.0
-- The CXX compiler identification is Clang 14.0.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Check for working CXX compiler: /usr/bin/clang++
-- Check for working CXX compiler: /usr/bin/clang++ - broken
CMake Error at /usr/share/cmake-3.22/Modules/CMakeTestCXXCompiler.cmake:62 (message):
  The C++ compiler

    "/usr/bin/clang++"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/user/Documents/OpenSpiel folders/open_spiel/build/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/usr/bin/gmake -f Makefile cmTC_da329/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_da329.dir/build.make CMakeFiles/cmTC_da329.dir/build
    gmake[1]: Entering directory '/home/user/Documents/OpenSpiel folders/open_spiel/build/CMakeFiles/CMakeTmp'
    Building CXX object CMakeFiles/cmTC_da329.dir/testCXXCompiler.cxx.o
    /usr/bin/clang++    -MD -MT CMakeFiles/cmTC_da329.dir/testCXXCompiler.cxx.o -MF CMakeFiles/cmTC_da329.dir/testCXXCompiler.cxx.o.d -o CMakeFiles/cmTC_da329.dir/testCXXCompiler.cxx.o -c "/home/user/Documents/OpenSpiel folders/open_spiel/build/CMakeFiles/CMakeTmp/testCXXCompiler.cxx"
    Linking CXX executable cmTC_da329
    /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_da329.dir/link.txt --verbose=1
    /usr/bin/clang++ CMakeFiles/cmTC_da329.dir/testCXXCompiler.cxx.o -o cmTC_da329 
    /usr/bin/ld: cannot find -lstdc++: No such file or directory
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    gmake[1]: *** [CMakeFiles/cmTC_da329.dir/build.make:100: cmTC_da329] Error 1
    gmake[1]: Leaving directory '/home/user/Documents/OpenSpiel folders/open_spiel/build/CMakeFiles/CMakeTmp'
    gmake: *** [Makefile:127: cmTC_da329/fast] Error 2
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:6 (project)
_______________________________________________________________________________________________________________________________________

Below are informations about Cuda version on my machine:
cat /usr/local/cuda/version.json

{
   "cuda" : {
      "name" : "CUDA SDK",
      "version" : "12.1.20230222"
   },
   "cuda_cccl" : {
      "name" : "CUDA C++ Core Compute Libraries",
      "version" : "12.1.55"
   },
   "cuda_cudart" : {
      "name" : "CUDA Runtime (cudart)",
      "version" : "12.1.55"
   },
   "cuda_cuobjdump" : {
      "name" : "cuobjdump",
      "version" : "12.1.55"
   },
   "cuda_cupti" : {
      "name" : "CUPTI",
      "version" : "12.1.62"
   },
   "cuda_cuxxfilt" : {
      "name" : "CUDA cu++ filt",
      "version" : "12.1.55"
   },
   "cuda_demo_suite" : {
      "name" : "CUDA Demo Suite",
      "version" : "12.1.55"
   },
   "cuda_gdb" : {
      "name" : "CUDA GDB",
      "version" : "12.1.55"
   },
   "cuda_nsight" : {
      "name" : "Nsight Eclipse Plugins",
      "version" : "12.1.55"
   },
   "cuda_nvcc" : {
      "name" : "CUDA NVCC",
      "version" : "12.1.66"
   },
   "cuda_nvdisasm" : {
      "name" : "CUDA nvdisasm",
      "version" : "12.1.55"
   },
   "cuda_nvml_dev" : {
      "name" : "CUDA NVML Headers",
      "version" : "12.1.55"
   },
   "cuda_nvprof" : {
      "name" : "CUDA nvprof",
      "version" : "12.1.55"
   },
   "cuda_nvprune" : {
      "name" : "CUDA nvprune",
      "version" : "12.1.55"
   },
   "cuda_nvrtc" : {
      "name" : "CUDA NVRTC",
      "version" : "12.1.55"
   },
   "cuda_nvtx" : {
      "name" : "CUDA NVTX",
      "version" : "12.1.66"
   },
   "cuda_nvvp" : {
      "name" : "CUDA NVVP",
      "version" : "12.1.55"
   },
   "cuda_opencl" : {
      "name" : "CUDA OpenCL",
      "version" : "12.1.56"
   },
   "cuda_sanitizer_api" : {
      "name" : "CUDA Compute Sanitizer API",
      "version" : "12.1.55"
   },
   "libcublas" : {
      "name" : "CUDA cuBLAS",
      "version" : "12.1.0.26"
   },
   "libcufft" : {
      "name" : "CUDA cuFFT",
      "version" : "11.0.2.4"
   },
   "libcufile" : {
      "name" : "GPUDirect Storage (cufile)",
      "version" : "1.6.0.25"
   },
   "libcurand" : {
      "name" : "CUDA cuRAND",
      "version" : "10.3.2.56"
   },
   "libcusolver" : {
      "name" : "CUDA cuSOLVER",
      "version" : "11.4.4.55"
   },
   "libcusparse" : {
      "name" : "CUDA cuSPARSE",
      "version" : "12.0.2.55"
   },
   "libnpp" : {
      "name" : "CUDA NPP",
      "version" : "12.0.2.50"
   },
   "libnvjitlink" : {
      "name" : "JIT Linker Library",
      "version" : "12.1.55"
   },
   "libnvjpeg" : {
      "name" : "CUDA nvJPEG",
      "version" : "12.1.0.39"
   },
   "libnvvm_samples" : {
      "name" : "NVVM Samples",
      "version" : "12.1.55"
   },
   "nsight_compute" : {
      "name" : "Nsight Compute",
      "version" : "2023.1.0.15"
   },
   "nsight_systems" : {
      "name" : "Nsight Systems",
      "version" : "2023.1.2.43"
   },
   "nvidia_driver" : {
      "name" : "NVIDIA Linux Driver",
      "version" : "530.30.02"
   },
   "nvidia_fs" : {
      "name" : "NVIDIA file-system",
      "version" : "2.15.1"
   }
}

_______________________________________________________________________________________________________________________________________

Also for your information:
nvidia-smi
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 390.157                Driver Version: 530.30.02                 |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:01:00.0 Off |                  N/A |
|  0%   29C    P8    16W / 340W |    425MiB / 10240MiB |      7%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|

**Comment by lanctot** [comment, marked as answer]
This does not seem unrelated to Cuda, seems like you are missing the standard C++ library:
```
 /usr/bin/ld: cannot find -lstdc++: No such file or directory
```

It should be fixable by just installing a few packages via apt. For more info, take a look at this: 
https://stackoverflow.com/questions/62612405/usr-bin-ld-cannot-find-lc

Try `sudo apt-get install gcc g++`, which should also install `libstdc++`. 

BTW I noticed you're using Python 3.11. We don't yet support officially Python 3.11 yet, but it's in the works (see https://github.com/deepmind/open_spiel/pull/1025) and will be available after Ubuntu 23.04 released. I personally tested it on Linux though and most things seem work already.
